## Exercise 4.4 Data Wrangling & Subsetting

    -Create a subset of data set
    -Create and use data dictionary
    -Export data as csv
  

#import libraries

In [4]:

import pandas as pd
import numpy as np
import os as os

/* Create path (to project file), import orders & products csv files */


In [2]:
path = r'/Users/margoshogren/Desktop/1 Margo-CF Folder/4 Python/4-2023 Instacart Basket Analysis'
df_ords = pd.read_csv(os.path.join(path,'2 Data','original data','orders.csv'),index_col = False)
df_prods = pd.read_csv(os.path.join(path,'2 Data','original data','products.csv'),index_col = False)

#view the order file, to see whats in it

In [3]:
df_ords.shape

(3421083, 7)

In [19]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

/*Exercise to drop column- eval_set */
## Use df.drop()

In [21]:

df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

# frequency count of 'day_since_prior_order' 

# df['col'].value_counts(dropna = False)

## (dropna = False) means to show the msg values

In [22]:

df_ords['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

## Observation of frequency of days since last order, is that there are 206209 missing values, the highest is a 30 count of 369363 and lowest is 26 count of 19016 

# Rename column order_dow to orders_day_of_week,
/*df.rename('col'= 'oldcol':'newcol'), inplace = True
inplace = True (means to overwrite) */

In [24]:
df_ords.rename(columns = {'order_dow':'orders_day_of_week'}, inplace = True)

In [25]:
df_ords.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

/* change type for column order_id from int to string */

#describe will only show the int types

In [27]:
df_ords['order_id'] = df_ords['order_id'].astype('str')
df_ords.describe()

user_id  order_number  orders_day_of_week  order_hour_of_day  \
count  3.421083e+06  3.421083e+06        3.421083e+06       3.421083e+06   
mean   1.029782e+05  1.715486e+01        2.776219e+00       1.345202e+01   
std    5.953372e+04  1.773316e+01        2.046829e+00       4.226088e+00   
min    1.000000e+00  1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.139400e+04  5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.026890e+05  1.100000e+01        3.000000e+00       1.300000e+01   
75%    1.543850e+05  2.300000e+01        5.000000e+00       1.600000e+01   
max    2.062090e+05  1.000000e+02        6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

/* change type of order_id, (Note: pandas calls the string an object) */

In [31]:
df_ords['order_id'].dtype

dtype('O')

## Import the department csv file; index_col = False because it already has an index

In [6]:
df_dep = pd.read_csv(os.path.join(path,'2 Data','original data','departments.csv'),index_col = False)

In [33]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

# We will transpose the above wide dataframe and make a data dictionary 
    -transpose df
    -reset the index
    -add a new header at the iloc '0' position
    -copy the df starting at 1
    -rename cols to new header  

## Transpose df_dep (department file) and call it
       df_t = df.T

In [35]:
df_dep_t = df_dep.T
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

## Add index, starting over at 0
    df_t.reset_index()

In [36]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

## Add new header at 0 index
    new_header = df_t.iloc[0]

In [ ]:
new_header = df_dep_t.iloc[0]

## Copy df starting at index 1 and call it
df_t_new = df_t [1:]

In [ ]:
df_dep_t_new = df_dep_t[1:]

In [39]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

## Add new header as the columns
    df_t_new.columns = new_header

In [ ]:
df_dep_t_new.columns = new_header
df_dep_t_new

In [41]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

## Create data dictionary with index 
data_dict = df_t_new.to_dict('index')  #Note .to_dict puts column in dictionary format

In [42]:
data_dict = df_dep_t_new.to_dict('index')

In [43]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [44]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

## Print out the element of 19 from data dictionary 
    print(data_dict.get('19'))

In [46]:
print(data_dict.get('19'))

{'department': 'snacks'}


## method 1 to get subset with dept id =19, using df[df]
   df_sub = df[df['col looking in]==19]                     
        -Note: bracket within df again to get only true values

In [ ]:
df_snacks = df_prods[df_prods['department_id']==19] 

In [52]:
df_snacks

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [53]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

## method 2 to get subset with dept id =19 (using loc)
    df_sub = df.loc[df['col lookin in']==19]

In [ ]:
df_snacks = df_prods.loc[df_prods['department_id']==19]

In [55]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

## method 3 to get subset with dept id =19, using .loc and .isin ([ ]); .isin is for a series 
    df_sub = df.loc[df['col looking in'].isin([19])]

In [ ]:
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

##  Question#2 -change a data type where needed 
    df['col to change'] = df['col to change].astype('new type')
    
    #(2)Change user_id data type to string, since it is not a variable that would use math functions

###Check the current data type for user_id

In [62]:
df_ords['user_id'].dtype

dtype('int64')

In [ ]:
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [65]:
df_ords['user_id']

0               1
1               1
2               1
3               1
4               1
            ...  
3421078    206209
3421079    206209
3421080    206209
3421081    206209
3421082    206209
Name: user_id, Length: 3421083, dtype: object

## Question#(3) Rename a column, with no overwrite 
    df.rename(columns = {'old name':'new name'})

In [68]:
df_ords.rename(columns= {'orders_day_of_week':'orders_starting_Sat_as_day1'})

order_id user_id eval_set  order_number  orders_starting_Sat_as_day1  \
0        2539329       1    prior             1                            2   
1        2398795       1    prior             2                            3   
2         473747       1    prior             3                            3   
3        2254736       1    prior             4                            4   
4         431534       1    prior             5                            4   
...          ...     ...      ...           ...                          ...   
3421078  2266710  206209    prior            10                            5   
3421079  1854736  206209    prior            11                            4   
3421080   626363  206209    prior            12                            1   
3421081  2977660  206209    prior            13                            1   
3421082   272231  206209    train            14                            6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

## Question#(4) what is the busiest hour for orders? Use a frequency count, 
    -Note: dropna = False means to show the missing values
    df['col'].value_counts(dropna = False)

### Answer (4): 10am is the busiest hour for orders.

In [71]:
df_ords['order_hour_of_day'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

## Question(5) the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary. 
    data_dict = df.to_dict('index')

In [183]:
data_dict = df_dep_t_new.to_dict ('index')
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [189]:
print(data_dict.get('4'))

{'department': 'produce'}


### Answer(5): The department dictionary shows 'produce' at index 4.

## Question(6) breakfast item sales. Create a subset containing only the required information.
    #df_sub = df.loc[df['department_id'].isin([14])]


In [174]:
df_prods_brkfst = df_prods.loc[df_prods['department_id'].isin([14])]

In [175]:
df_prods_brkfst.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

# Question(7- part1) Find the department no.s for dinner party for (alcohol, deli, beverages, meat seafood)
    df_sub = df.loc['col to look in'].isin(['element1',element2'])]

In [119]:

df_dep_t_new_dnrpty = df_dep_t_new.loc[df_dep_t_new['department'].isin(['alcohol', 'deli','beverages','meat seafood'])]
df_dep_t_new_dnrpty

department_id    department
5                   alcohol
7                 beverages
12             meat seafood
20                     deli

## Question(7 part-2) Your task is to find all observations from the entire dataframe that include items from the following departments:
    ###alcohol, deli, beverages, meat, seafood.
    df_dep_dnrpty = df_dep.loc[df_dep['department'].isin(['alcohol', 'deli','beverages','meat seafood'])]

In [7]:
df_prods.shape

(49693, 5)

In [107]:
df_prods_dnrprty = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

In [109]:
df_prods_dnrprty.head()

product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
16          17                               Rendered Duck Fat        35   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1

## Question(8) How many rows does the last dataframe you created have?
    ## Use df.shape
    
    #Answer 7650

In [111]:
df_prods_dnrprty.shape

(7650, 5)

## Question(9) customer with a "user_id" of “1.” Extract all the information you can about this user.
    df_info = df[df['col'] == 'element']

In [114]:
df_ords_usr1 = df_ords[df_ords['user_id']== '1']

In [115]:
df_ords_usr1

order_id user_id eval_set  order_number  orders_day_of_week  \
0   2539329       1    prior             1                   2   
1   2398795       1    prior             2                   3   
2    473747       1    prior             3                   3   
3   2254736       1    prior             4                   4   
4    431534       1    prior             5                   4   
5   3367565       1    prior             6                   2   
6    550135       1    prior             7                   1   
7   3108588       1    prior             8                   1   
8   2295261       1    prior             9                   1   
9   2550362       1    prior            10                   4   
10  1187899       1    train            11                   4   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0

## Question(10) What basic stats can you provide based on the information you have?
    Use df.describe()

In [116]:
df_ords_usr1.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

### Observations from above description of User1. 
User1 has 11 orders, on average orders at 10am on Sunday. All orders were between Saturday(1) and Tuesday(4) and average 19 days between orders, with a minimun of 0 days apart and maximum of 30 days apart.

### Question(11) Included Section headers and comments

## Question (12) Export as orders_wrangled.csv to prepared data
        df.to_csv(os.path.join(path,'path','file_ren.csv'))

In [191]:
df_ords.to_csv(os.path.join(path,'2 Data','prepared data','orders_wrangled.csv'))

## Question(13) Export the df_dep_t_new dataframe as “departments_wrangled.csv” to prepared data folder 
        df.to_csv(os.path.join(path,'file_ren.csv'))

In [192]:
df_dep_t_new.to_csv(os.path.join(path,'2 Data','prepared data','departments_wrangled.csv'))

#### End of Exercise 4.4